in_toto/verifylib.go

```diff
@@ -0,0 +1,502 @@
+  }
+  // Uses timesone of expires, i.e. UTC
+  if time.Until(expires) < 0 {
+    return fmt.Errorf("Layout has expired on '%s'.", expires)
+  }
+  return nil
+}
+
+
+func VerifyLayoutSignatures(layoutMb Metablock, layoutKeys map[string]Key) error {
+  if len(layoutKeys) < 1 {
+    return fmt.Errorf("Layout verification requires at least one key.")
+  }
+
+  for _, key := range layoutKeys {
+    if err := layoutMb.VerifySignature(key); err != nil {
+      return err
+    }
+  }
+  return nil
+}
+
+
+func InTotoVerify(layoutPath string, layoutKeys map[string]Key, linkDir string) error {
+
+  var layoutMb Metablock
+
+  // Load layout
+  if err := layoutMb.Load(layoutPath); err != nil {
+    return err
+  }
+
+  // Verify root signatures
+  if err := VerifyLayoutSignatures(layoutMb, layoutKeys); err != nil {
+    return err
+  }
+
+  // Extract the layout from its Metablock container (for further processing)
+  layout := layoutMb.Signed.(Layout)
+
+  // Verify layout expiration
+  if err := VerifyLayoutExpiration(layout); err != nil {
+    return err
+  }
+
+  // Substitute parameters
+  // TODO
+
+  // Load links for layout
+  stepsMetadata, err := LoadLinksForLayout(layout, linkDir)
+  if err != nil {
+    return err
+  }
+
+  // Verify link signatures
+  stepsMetadataVerified, err := VerifyLinkSignatureThesholds(layout, stepsMetadata)
+  if err != nil {
+    return err
+  }
+
+  // Verify sublayouts
+  // TODO
+
+  // Verify command alignment (WARNING only)
+  VerifyStepCommandAlignment(layout, stepsMetadataVerified)
+
+  // Given that signature thresholds have been checked above and the rest of
+  // the relevant link properties, i.e. materials and products, have to be
+  // exactly equal, we can reduce the map of steps metadata. However, we error
+  // if the relevant properties are not equal among links of a step.
+  stepsMetadataReduced, err := ReduceStepsMetadata(layout, stepsMetadataVerified)
+  if err != nil {
+    return err
+  }
+
+  // Go does not allow to pass []Step as []interface{}
+  // We have to manually copy first :(
+  // https://golang.org/doc/faq#convert_slice_of_interface
+  stepsI := make([]interface{}, len(layout.Steps))
+  for i, v := range layout.Steps {
+      stepsI[i] = v
+  }
+  // Verify artifact rules
+  if err := VerifyArtifacts(stepsI, stepsMetadataReduced); err != nil {
+    return err
+  }
+
+  inspectionMetadata, err := RunInspections(layout)
+  if err != nil {
+    return err
+  }
+
+  // Add steps metadata to inspection metadata, because inspection artifact
+  // rules may also refer to artifacts reported by step links
+  for k, v := range stepsMetadataReduced {
+    inspectionMetadata[k] = v
+  }
+
+  // Convert []Inspections to []interace{} (see `stepsI` above)
+  inspectionsI := make([]interface{}, len(layout.Inspect))
+  for i, v := range layout.Inspect {
+      inspectionsI[i] = v
+  }
+
+  if err := VerifyArtifacts(inspectionsI, inspectionMetadata); err != nil {
+    return err
+  }
+
+  return nil
```

Comment:
Another `if`/ implicit `else` / actually returning `err`.

Reply:
See https://github.com/in-toto/in-toto-golang/pull/1#discussion_r230004116